ops/model.py

```diff
@@ -1324,7 +1368,227 @@ def list_files(self, path: str, *, pattern: str = None,
+                            make_dirs=True,
+                            permissions=f.permissions,
+                            user_id=f.user_id,
+                            user=f.user,
+                            group_id=f.group_id,
+                            group=f.group)
+            except Exception as err:
+                errors.append((source_path, err))
+        if len(errors) > 0:
+            raise MultiPushPullError('failed to push one or more files', errors)
+
+    def pull_path(self,
+                  source_path: typing.Union[StrOrPath, typing.List[StrOrPath]],
+                  dest_dir: StrOrPath):
+        """Recursively pull a remote path or files to the local system.
+
+        Pulling is attempted to completion even if errors occur during the process.  All errors are
+        collected incrementally. After copying has completed, if any errors occurred, a single
+        MultiPushPullError is raised containing details for each error.
+
+        Assuming the following files exist remotely:
+
+            * /foo/bar/baz.txt
+            * /foo/foobar.txt
+            * /quux.txt
+
+        You could pull the following ways:
+
+            # copy one file
+            container.pull_path('/foo/foobar.txt', '/dst')
+            # Destination results: /dst/foobar.txt
+
+            # copy a directory
+            container.pull_path('/foo', '/dst')
+            # Destination results: /dst/foo/bar/baz.txt, /dst/foo/foobar.txt
+
+            # copy a directory's contents
+            container.pull_path('/foo/*', '/dst')
+            # Destination results: /dst/bar/baz.txt, /dst/foobar.txt
+
+            # copy multiple files
+            container.pull_path(['/foo/bar/baz.txt', 'quux.txt'], '/dst')
+            # Destination results: /dst/baz.txt, /dst/quux.txt
+
+            # copy a file and a directory
+            container.pull_path(['/foo/bar', '/quux.txt'], '/dst')
+            # Destination results: /dst/bar/baz.txt, /dst/quux.txt
+
+        Args:
+            source_path: A single path or list of paths to pull from the remote system.  The
+                paths can be either a file or a directory but must be absolute paths.  If
+                source_path is a directory, the directory base name is attached to the destination
+                directory - i.e.  the source path directory is placed inside the destination
+                directory.  If a source path ends with a trailing "/*" it will have its *contents*
+                placed inside the destination directory.
+            dest_dir: Local destination directory inside which the source dir/files will be
+                placed.
+        """
+        if os.name == 'nt':
+            raise RuntimeError('Container.pull_path is not supported on Windows-based systems')
+
+        source_paths = source_path if isinstance(source_path, list) else [source_path]
+        source_paths = [str(p) for p in source_paths]
+        dest_dir = str(dest_dir)
```

Comment:
If `pathlib` were used, this would mostly not be necessary ;)

Broadly, I could say that I wish Python type aliases allowed something more like Go's interfaces so you could do the "right thing" without manually jamming it in, but this could be "condensed" (golfed, just pretty readable golfing) as:

```python
source_paths, dest_dir = map(lambda x: str(x), source_path if isinstance(source_path, list) else [source_path]), str(dest_dir)
```

"Rationalizing" the arguments isn't that terrible here. From my POV, twice doesn't quite hit the threshold for breaking it out (the definition/call/docstring plus the actual function and added returns) will end up being bulkier than it is otherwise, and there don't appear to be enough other cases to really make use of it. There are 10 in `ops.testing`, but those mostly pass it through without modification to other places which pass it into `Path` anyway

Reply:
I think it's short enough and only repeated twice - so I'm leaning toward leaving it alone.  Code should be all updated to using pathlib as much as reasonable now (I didn't push it into the pebble client methods - maybe for another day).